feat: add fabric option

troZee opened the pull request:
### Summary

Close https://github.com/callstack/react-native-builder-bob/issues/238 

<img width="554" alt="Screenshot 2022-09-06 at 15 25 48" src="https://user-images.githubusercontent.com/12766071/188647061-216036d8-3df3-40a0-a478-e4e7a5f46a03.png">

### Test plan

- [ ] Turbo module with backward compat (experimental)
- [ ] Turbo module (experimental)
- [ ]  Native module
- [ ]  Native view
- [ ] JavaScript library

troZee commented on packages/create-react-native-library/src/index.ts:
```diff
@@ -80,7 +89,13 @@ type Answers = {
-  type?: 'module-legacy' | 'module-turbo' | 'module-mixed' | 'view' | 'library';
+  type?:
+    | 'module-legacy'
+    | 'module-new-arch'
```
@satya164 , I changed 'xxx-turbo' => 'xxx-new-arch'  to avoid any confusion. 
Fabric will refer to view, turbomodule will refer to module. Since the fact fabric is a new renderer, I would like to use `new-arch` postfix for the new arch

troZee commented on packages/create-react-native-library/src/index.ts:
```diff
@@ -365,10 +391,12 @@ async function create(argv: yargs.Arguments<any>) {
       architecture,
       turbomodule,
+      fabricView,
+      newArchitecture,
```
Do we need `newArchitecture` ? @MateWW . `architecture` indicates, what arch is it, so IMO this is redundant

troZee commented on packages/create-react-native-library/templates/common/$package.json:
```diff
@@ -53,7 +53,7 @@
     "@types/jest": "^28.1.2",
     "@types/react": "~17.0.21",
-    "@types/react-native": "0.68.0",
+    "@types/react-native": <%-/* TODO: bump version of rn to 0.69.0 */ project.newArchitecture ? '"0.69.5"':'"0.68.0"'-%>,
```
Will adjust this todo after merge https://github.com/callstack/react-native-builder-bob/pull/271

troZee commented on packages/create-react-native-library/templates/example/example/android/app/src/main/jni/Android.mk:
```diff
@@ -36,8 +36,8 @@ LOCAL_SHARED_LIBRARIES := \
-<% if (project.turbomodule) { -%>
-  libreact_codegen_<%- project.name -%> \
+<% if (project.newArchitecture) { -%>
+  libreact_codegen_<%- project.view ? project.name+'View': project.name -%> \
```
It will not be needed in RN 70

troZee commented on packages/create-react-native-library/templates/example/example/android/app/src/main/jni/MainComponentsRegistry.cpp:
```diff
@@ -4,6 +4,9 @@
 #include <fbjni/fbjni.h>
 #include <react/renderer/componentregistry/ComponentDescriptorProviderRegistry.h>
 #include <react/renderer/components/rncore/ComponentDescriptors.h>
+<% if (project.fabricView) { -%>
```
ditto

troZee commented on packages/create-react-native-library/templates/java-view-mixed-library/android/src/legacy/{%- project.name %}ViewManager.java:
```diff
@@ -0,0 +1,61 @@
+    super.receiveCommand(view, commandId, args);
+    String color = args.getString(0);
+
+    switch (commandId) {
```
@MateWW should I add command to spec ?

MateWW replied:
Yeah lets copy spec from my unicorn example

troZee commented on packages/create-react-native-library/templates/common/README.md:
````diff
@@ -18,7 +18,7 @@ import { <%- project.name -%>View } from "<%- project.slug -%>";
 <<%- project.name -%>View color="tomato" />
 ```
-<% } else if (project.architecture == 'turbo') { -%>
+<% } else if (project.architecture == 'new_arch') { -%>
````
Since the fact, naming convention is ambiguous, I reverted my changes. Refactor will be part of another discussion

troZee commented:
@MateWW @satya164 @atlj 
<img width="352" alt="rename" src="https://user-images.githubusercontent.com/12766071/189102831-7e27f4c2-b065-4112-8a0b-45d22237e7f1.png">

I renamed all files to keep this name convenction `${language}-${type}-${architecture}`. Some folders have not been changed, bc it would not make any sense eg `kotlin-library`, since template for both arch is the same. 

Moreover I removed `turbomodule, fabricView, newArchitecture` properties to avoid any confusion and use only needed ones. Templates has been also updated with the new conditions. If those conditions are not clear, please let me know. 

Right now, I am waiting for https://github.com/callstack/react-native-builder-bob/pull/271 to be merged and then I will rebase my branch with the main branch and fix the changes.

troZee commented:
> Does it make sense to replace "new" with "fabric" and "turbomodule" for views and modules respectively?

I would prefer to stick into that template `${language}-${type}-${architecture}`, since it would be much easier to parse those dirs based on chosen options. Moreover we can assume, that fabric is view plus new-arch and turbomodule is module plus new arch. Changing names to turbomodule or fabric would add additional complexity and can be ambiguous in case of mixed (backward compatibility) type

thymikee commented:
My point is that "new" is new today, but may be old in 3 years 😅 but this is the naming that RN uses as well so I won't protest too much

atlj commented:
There is a bug and this is probably a regression after 881aaf4

## Issue

```sh
 ./create-react-native-library test      ✘ PIPE 5s
✔ What is the name of the npm package? … react-native-test
✔ What is the description for the package? … sdf
✔ What is the name of package author? … Burak
✔ What is the email address for the package author? … burak.guner@callstack.com
✔ What is the URL for the package author? … https://
✔ What is the URL for the repository? … https://
✔ What type of library do you want to develop? › Native fabric view with backward compat (experimental)
create-react-native-library <name>

create a react native library

Options:
  --help          Show help                                            [boolean]
  --version       Show version number                                  [boolean]
  --slug          Name of the npm package                               [string]
  --description   Description of the npm package                        [string]
  --author-name   Name of the package author                            [string]
  --author-email  Email address of the package author                   [string]
  --author-url    URL for the package author                            [string]
  --repo-url      URL for the repository                                [string]
  --languages     Languages you want to use
      [choices: "java-objc", "java-swift", "kotlin-objc", "kotlin-swift", "cpp",
                                                                           "js"]
  --type          Type of library you want to develop[choices: "module", "view"]
  --example       Type of example app                [choices: "expo", "native"]

[Error: ENOENT: no such file or directory, scandir '/Users/burakguner/kod/react-native-builder-bob/packages/create-react-native-library/templates/java-view-mixed-library'] {
  errno: -2,
  code: 'ENOENT',
  syscall: 'scandir',
  path: '/Users/burakguner/kod/react-native-builder-bob/packages/create-react-native-library/templates/java-view-mixed-library'
}
```

## Steps to reproduce

1. run `create-react-native-library` 
2.  choose `Native fabric view with backward compat (experimental)` for `What type of library do you want to develop?`

atlj commented on packages/create-react-native-library/src/index.ts:
```diff
@@ -17,43 +18,46 @@ const COMMON_FILES = path.resolve(__dirname, '../templates/common');
   module_mixed: path.resolve(__dirname, '../templates/java-library-mixed'),
-  view: path.resolve(__dirname, '../templates/java-view-library'),
+  view_legacy: path.resolve(__dirname, '../templates/java-view-legacy'),
+  view_mixed: path.resolve(__dirname, '../templates/java-view-mixed-library'),
```
I couldn't see this folder under templates and this might be responsible for [this issue](https://github.com/callstack/react-native-builder-bob/pull/273#issuecomment-1241710146)

troZee replied:
Thank you for checking it. That was an issue

troZee commented on packages/create-react-native-library/templates/common/$package.json:
```diff
@@ -53,7 +53,7 @@
     "@types/jest": "^28.1.2",
     "@types/react": "~17.0.21",
-    "@types/react-native": "0.68.0",
+    "@types/react-native": <%- project.architecture === "new" ? '"0.69.6"':'"0.68.0"' -%>,
```
https://www.npmjs.com/package/@types/react-native please change it to 0.70

troZee commented on packages/create-react-native-library/src/index.ts:
```diff
@@ -19,45 +19,50 @@ const COMMON_FILES = path.resolve(__dirname, '../templates/common');
   module_mixed: path.resolve(__dirname, '../templates/kotlin-library-mixed'),
-  view: path.resolve(__dirname, '../templates/kotlin-view-library'),
+  view_legacy: path.resolve(__dirname, '../templates/kotlin-view-legacy'),
+  // MISSING FILES
```
@satya164 I added those two keys `view_mixed` and `view_new` to conform to `JAVA_FILES` shape. Ofc those templates are not implemented yet. Should I remove it ?

satya164 replied:
@troZee if not implemented yet imo better not to add them here or it can cause confusion

satya164 commented on packages/create-react-native-library/templates/common/CONTRIBUTING.md:
```diff
@@ -94,6 +94,7 @@ To edit the Objective-C or Swift files, open `example/ios/<%- project.name -%>Ex
 To edit the Java or Kotlin files, open `example/android` in Android studio and find the source files at `<%- project.identifier -%>` under `Android`.
 
 <% } -%>
+
```
need to remove this new line to make prettier happy for generated files. in vs code you can choose "Save without formatting" in command palette if you have auto format turned on.

satya164 commented on packages/create-react-native-library/templates/common/README.md:
````diff
@@ -11,29 +11,35 @@ npm install <%- project.slug %>
 const result = await multiply(3, 7);
 ```
+
 <% } -%>
````
Avoid adding new lines here

satya164 commented on packages/create-react-native-library/templates/objc-view-new/ios/{%- project.name %}View.mm:
```diff
@@ -0,0 +1,80 @@
+    return <%- project.name -%>View.class;
+}
+
+- hexStringToColor:(NSString *)stringToConvert
+{
+    NSString *noHashString = [stringToConvert stringByReplacingOccurrencesOfString:@"#" withString:@""];
+    NSScanner *stringScanner = [NSScanner scannerWithString:noHashString];
+    
+    unsigned hex;
+    if (![stringScanner scanHexInt:&hex]) return nil;
+    int r = (hex >> 16) & 0xFF;
+    int g = (hex >> 8) & 0xFF;
+    int b = (hex) & 0xFF;
+    
+    return [UIColor colorWithRed:r / 255.0f green:g / 255.0f blue:b / 255.0f alpha:1.0f];
+}
+
+- (void)changeBackgroundColor:(nonnull NSString *)color {
+    [_view setBackgroundColor:[self hexStringToColor:color]];
+}
```
I'd prefer to have simpler code in these samples, so maybe something like increasing border width or something which doesn't require color generation and conversion?

satya164 commented on packages/create-react-native-library/templates/objc-view-mixed/ios/{%- project.name %}View.mm:
```diff
@@ -0,0 +1,82 @@
+    return [UIColor colorWithRed:r / 255.0f green:g / 255.0f blue:b / 255.0f alpha:1.0f];
+}
+
+- (void)changeBackgroundColor:(nonnull NSString *)color {
+    [_view setBackgroundColor:[self hexStringToColor:color]];
+}
```
Is it necessary to include these? I usually try to keep these samples super basic since they are supposed to be a starting point - usually with a very basic example API that shows that it's working. This way it's easy to delete that example and add what I need.

I'd avoid including API examples when not essential since otherwise, it's harder to figure out what can be safely removed when implementing my own module. And maybe a comment with link to official docs where they can read about APIs they can use.

For views, I'd include a very basic view that accepts one prop with any essential scaffolding, and remove example APIs that show other usage.

satya164 commented on packages/create-react-native-library/templates/objc-view-mixed/ios/{%- project.name %}ViewManager.mm:
```diff
@@ -0,0 +1,45 @@
+  [view setBackgroundColor:[self hexStringToColor:json]];
+}
+
+- hexStringToColor:(NSString *)stringToConvert
```
Do both the viewmanager and view need to include these code?

troZee commented on .github/workflows/build-templates.yml:
```diff
@@ -117,7 +117,7 @@ jobs:
         working-directory: ${{ env.work_dir }}
         run: |
-          yarn lint
+          yarn lint --fix
```
Remove this before merge

migueldaipre commented:
@satya164 Any news when this will be released?

thymikee commented:
🔜
